lua/compe_tabnine/init.lua

```diff
@@ -0,0 +1,182 @@
+local compe = require'compe'
+local api = vim.api
+local fn = vim.fn
+--
+
+local function dump(...)
+    local objects = vim.tbl_map(vim.inspect, {...})
+    print(unpack(objects))
+end
+
+local function json_decode(data)
+  local status, result = pcall(vim.fn.json_decode, data)
+  if status then
+    return result
+  else
+    return nil, result
+  end
+end
+
+-- locate the binary here, as expand is relative to the calling script name
+local binary = nil
+if fn.has("mac") == 1 then
+	binary = fn.expand("<sfile>:p:h:h:h") .. "/binaries/TabNine_Darwin"
+elseif fn.has('unix') == 1 then
+	binary = fn.expand("<sfile>:p:h:h:h") .. "/binaries/TabNine_Linux"
+else
+	binary = fn.expand("<sfile>:p:h:h:h") .. "/binaries/TabNine_Windows"
+end
+
+
+local Source = {
+	items = {};
+	max_lines = 1000;
+	max_num_results = 20;
+	last_initiated = 0;
+	last_finished = 0;
+}
+
+--- get_metadata
+function Source.get_metadata(_)
+	return {
+		priority = 5000;
+		dup = 0;
+		menu = '[TN]';
+		-- by default, do not sort, as TabNine already sorts by relevance
+		sort = false;
+	}
+end
+
+--- determine
+function Source.determine(_, context)
+  return compe.helper.determine(context)
+end
+
+
+Source._do_complete = function()
+	if Source.job == 0 then
+		return
+	end
+
+	Source.last_initiated = Source.last_initiated + 1
+	Source.items = {}
+
+	local cursor=api.nvim_win_get_cursor(0)
+	local cur_line = api.nvim_get_current_line()
+	local cur_line_before = string.sub(cur_line, 0, cursor[2])
+	local cur_line_after = string.sub(cur_line, cursor[2]+1) -- include current character
+
+	local region_includes_beginning = false
+	local region_includes_end = false
+	if cursor[1] - Source.max_lines <= 1 then region_includes_beginning = true end
+	if cursor[1] + Source.max_lines >= fn['line']('$') then region_includes_end = true end
+
+	local lines_before = api.nvim_buf_get_lines(0, cursor[1] - Source.max_lines , cursor[1]-1, false)
+	table.insert(lines_before, cur_line_before)
+	local before = fn.join(lines_before, "\n")
```

Comment:
I think this could be replaced with `table.concat(lines_before, "\n")`, we should try not to use `vim.fn` that often because that might cause some performance issue.

Reply:
Sure.
Any other Lua fixes will be appreciated as Im not that fluent in Lua.